### PR TITLE
Add `log_grad_norm` hook to `LightningModule`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #7631

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
great, already integrated in new loop!

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -440,6 +440,20 @@ def __sync(
     def __check_none(name: str, value: Any, _) -> Any:
         raise ValueError(f'`self.log({name}, {value})` was called, but `None` values cannot be logged')
 
+    def log_grad_norm(self, grad_norm_dict: Dict[str, torch.Tensor]) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:443]: is adding a new hook the only option here? his feels very strange that this only controls the log behavior, and not the calculation. what's the relationship between this and `on_after_bacwkard`?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -355,19 +351,15 @@ def __auto_reduce_results_on_epoch_end(self, epoch_output):
 
         return epoch_log_metrics, epoch_progress_bar_metrics
 
-    def log_train_step_metrics(self, batch_output):
+    def log_train_step_metrics(self):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:354]:
```suggestion
    def log_train_step_metrics(self) -> None:
```

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -440,6 +440,20 @@ def __sync(
...
+        Args:
+            grad_norm_dict: Dictionary containing current grad norm metrics
+
+        Example::
+
+            # DEFAULT
+            def log_grad_norm(self, grad_norm_dict):
+                self.log_dict(grad_norm_dict, on_step=False, on_epoch=True, prog_bar=False, logger=True)
+        """
+        self.log_dict(grad_norm_dict, on_step=True, on_epoch=True, prog_bar=True, logger=True)
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:455]: while not BC, is `prog_bar=False` a better default?

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:443]: Hey @ananthsub. Users wanted to customise how they aggregate the grad_norm and log them.
Here, they are free to change the reduce_fx to max, or even computed the standard deviation across batches if they want.

`on_after_bacwkard` is independent for this function.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:455]: @carmocca Any thoughts on this one ? I see value in both.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:354]: Done in #7631

**Review by SeanNaren — APPROVED** [review]

**Pull request merged**